### ISSUE TITLE
cURL Feedback

### ISSUE THREAD

**Issue by janbrummer** [janbrummer/libproxy2#55]
Hi @bagder,

in the past weeks I've had the chance to rewrite libproxy to address your valid & good open points in order to make sure it can be integrated into cURL. We have now reached the point where it would be really nice to give you a chance to have a look at it and ask for changes/feedback. 

This rewrite has been reworked closely with the current libproxy maintainer @DimStar77 and has had an initial reviewed by @mcatanzaro . 

One of the big open points is how we want to implement the async operations: https://github.com/janbrummer/libproxy2/pull/22

Thanks for your feedback o/

**Comment by mcatanzaro** [comment]
> One of the big open points is how we want to implement the async operations: https://github.com/janbrummer/libproxy2/pull/22

This is the most important point IMO, as I don't know what to do about it.

The other concerns from [your blog post](https://daniel.haxx.se/blog/2022/08/12/the-dream-of-auto-detecting-proxies/) are still mostly outstanding, but we know how to handle them:

> Improve the project with docs. We cannot safely rely on a library and its APIs if we don’t know exactly what to expect.

Docs definitely still require improvement, but we know how to do this.

> There needs to be at least basic tests that verify its functionality. We cannot fix and improve the library safely if we cannot check that it still works and behaves as expected. Tests is the only reliable way to make sure of this.

libproxy still has no tests, but we can start by coping and adjusting the tests from glib-networking, which a pretty decent. PAC will require additional testing.

> Add CI jobs that build the project and run those tests. To help the project better verify that things don’t break along the way.

Still needed (but maybe not before merging into upstream libproxy).

> Consider improving the API. To be fair, it is extremely simple now, but it’s also so simple that that it becomes ineffective and quirky in some use cases.

Besides adding async API, anything else you'd like to see?

> Allow for external URL parser and URL retrieving etc to avoid blocking, double-parsing and to reuse caches properly. It would be ridiculous for curl to use a library that has its own separate (semi-broken) HTTP transfer and its own (synchronous) name resolving process.

I think it's OK for libproxy to just use libcurl unconditionally, then we can avoid the need for adding this API. OK? (Technically curl is an implementation detail and it could get swapped out in the future for libsoup or something, but that seems pretty unlikely.)

**Comment by mcatanzaro** [comment]
> libproxy still has no tests, but we can start by coping and adjusting the tests from glib-networking, which a pretty decent. PAC will require additional testing.

Oh, I've been corrected, they're just under src/ but I didn't think to look under there. I would consider moving those up one directly.

>> Add CI jobs that build the project and run those tests. To help the project better verify that things don’t break along the way.
> Still needed (but maybe not before merging into upstream libproxy).

This was wrong too; I see there is working CI, but I just didn't notice.

**Comment by janbrummer** [comment]
FTR:: Docs are autogenerated and pushed to a GitHub page using CI as well

**Comment by bagder** [comment]
Where do I find the docs for the async API?

**Comment by janbrummer** [comment]
> Where do I find the docs for the async API?

That's what is linked in the first post: https://github.com/janbrummer/libproxy2/pull/22

It's still in progress and also the reason why we have involved you right now. You have the chance to get the implementation you would need.

**Comment by bagder** [comment]
It looks as if it uses both glib and threads? Is that correct? I hate to be a downer, but both of those disqualifies it as a contender for use in libcurl.

**Comment by janbrummer** [comment]
glib yes. What's your concern about glib? You are using other features in curl with different library requirement. I'm curious. What's the reason?

Is it just the linked library? Because for e.g. the GNOME configuration we do need the glib library.

**Comment by bagder** [comment]
It's not a veto from me, it's just that glib is itself going to be way larger than libcurl itself for just a small support feature. It seems... silly.

**Comment by bagder** [comment]
The event-loop needs to use glib to use the API?

**Comment by bagder** [comment]
Can you tell I miss "proper" documentation?

**Comment by mcatanzaro** [comment]
GLib is probably already a dependency of 80-90% of the apps on your desktop, so I can't imagine why avoiding it would be desired. Asking for a rewrite to avoid such an ubiquitous dependency doesn't seem reasonable. If building for an embedded device where you really don't want any GLib, just build curl without the libproxy support?

What's your concern regarding threads? The point of the async API is to hide any use of threads behind a non-threaded interface, so you don't have to know about the threads or use threads yourself to avoid blocking.

> The event-loop needs to use glib to use the API?

Yes, but only for the async API proposed in #22, which I don't think we want to land. That's the primary reason we're seeking your feedback now. I don't know what an async API without GLib would even look like because we have to have some sort of event loop integration to execute the async callback, and adding hooks for you to plug in an arbitrary event loop seems overkill for such a small library. Specific question for you now is: "what would curl like a libproxy async API to look like?"

**Comment by mcatanzaro** [comment]
@janbrummer can you link to any built documentation please? All I could find was [this page](https://github.com/janbrummer/libproxy2/blob/main/docs/build-howto.md) (which should be renamed because it looks like a description of how to build the library rather than how to use it).

**Comment by janbrummer** [comment]
> @janbrummer can you link to any built documentation please? All I could find was [this page](https://github.com/janbrummer/libproxy2/blob/main/docs/build-howto.md) (which should be renamed because it looks like a description of how to build the library rather than how to use it).
I guess you are not talking about the config-env-test, right?

**Comment by janbrummer** [comment]
Sent you the link in Chat.

@bagder Back to my question: What's your concern about glib? Should i arrange an online meeting to discuss this?

**Comment by janbrummer** [comment]
@bagder  Do you have a free slot tomorrow?

**Comment by janbrummer** [comment]
Another ping for @bagder . Please speak up.

**Comment by bagder** [comment]
You're free to do whatever you want in your project, you don't need my permission. I was mostly surprised that a supposedly small and "simple" project such as this would even consider using glib. My experience with projects taking the glib route has through times not been overly positive. But that can also have been bad luck or issues since fixed.

As a potential future user of the library I'm much more curious and interested in the updated API and design.

**Comment by janbrummer** [comment]
Thanks for your feedback. There are several reasons why we have integrated glib features: to name a few
 - Testing Framework
 - Make use of existing plugin loader instead of writing one of our own and introduce new bugs. No need to reinvent the wheel
 - Automatic documentation out of code
 - Gobject Introspection bindings which allows almost every language to make use of this lib instead of writing all on our own
 - It's already a requirement for GNOME configuration extraction
 - Simple D-Bus implementation as we have also added a D-Bus Service for Linux

As you have also asked for an async version we would like to take the chance to involve you here. So that it will be integrated in curl as well. See: https://github.com/janbrummer/libproxy2/pull/22

Feel free to raise a ticket in case you spot something we need to address. I'm closing this on. Thanks o/
